### DISCUSSION TITLE
dot (from graphviz package) much slower on M1 as on very old Intel CPU

### DISCUSSION THREAD

**Discussion by peteromueller** [Homebrew/discussions#1773] [Everyday usage]
Hello,

I noticed that calling dot on my new M1 is quite slow.

On my old Intel Mac (Core Duo from 2009) the test results in about 150ms execution time.
The installed dot version is 2.42.2. Btw. this is the time also measure on a FreeBSD box.

With my iMac M1 there is a noticeable delay. Call with the same dot file takes about 350-450ms.
Installed dot version is 2.47.3

The test I use:
gdate +%N; dot -Tpng -o test.png -Tjson -o test.json test.dot; gdate +%N

Is there something I can do? Otherwise I don't notice that my new M1 is slower. The opposite is the case. Installing and compiling brew packages is much much faster.

Any hint what I can try out is very welcome.

Thanks,
Peter

**Comment by peteromueller** [comment]
Does this help? Is the dependency to /usr/lib/libexpat.1.dylib correct? There no such file in the /usr/lib

```
otool -L dot /opt/homebrew/bin/dot
/opt/homebrew/bin/dot:
	/opt/homebrew/Cellar/graphviz/2.47.3/lib/libgvc.6.dylib (compatibility version 7.0.0, current version 7.0.0)
	/opt/homebrew/opt/libtool/lib/libltdl.7.dylib (compatibility version 11.0.0, current version 11.1.0)
	/opt/homebrew/Cellar/graphviz/2.47.3/lib/libxdot.4.dylib (compatibility version 5.0.0, current version 5.0.0)
	/opt/homebrew/Cellar/graphviz/2.47.3/lib/libcgraph.6.dylib (compatibility version 7.0.0, current version 7.0.0)
	/opt/homebrew/Cellar/graphviz/2.47.3/lib/libpathplan.4.dylib (compatibility version 5.0.0, current version 5.0.0)
	/usr/lib/libexpat.1.dylib (compatibility version 7.0.0, current version 8.0.0)
	/usr/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.11)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1292.100.5)
	/opt/homebrew/Cellar/graphviz/2.47.3/lib/libcdt.5.dylib (compatibility version 6.0.0, current version 6.0.0)
```

**Reply by carlocab** [reply]
That linkage shouldn't be a problem. On Big Sur, system libraries no longer live in the file system but instead can be found in the dynamic linker cache. As long as you can `dlopen` them then it shouldn't matter that you can't find the file.

You're probably better off reporting the poor performance to the `graphviz` developers to see if its reproducible and if they can do something about it.

**Reply by peteromueller** [reply]
Hello,

is there a way to build the older version 2.42.2 which I run before.
Maybe the issue is in the new version.

How can I install an older version?
brew install graphviz@2.42.2  did not work.
How do I find out which older versions are supported / available?

Thanks,
Peter

**Reply by gromgit** [reply]
> How can I install an older version?

The general procedure: `brew extract` the desired version of the formula into [your own tap](https://docs.brew.sh/How-to-Create-and-Maintain-a-Tap), then build it there.
```
brew tap-new peteromueller/versions
brew extract --version=2.42.2 graphviz peteromueller/versions
brew install peteromueller/versions/graphviz@2.42.2
```
However, v2.42.2 goes back far enough in Homebrew history that there have been incompatible changes in the `bottle` block syntax since then. Because Homebrew has to load the formula in order to version it, this happens:
```
$ brew extract --version=2.42.2 graphviz gromgit/dev
==> Searching repository history
Error: graphviz: Calling `sha256 "digest" => :tag` in a bottle block is disabled! Use `brew style --fix` on the formula to update the style or use `sha256 tag: "digest"` instead.
Please report this issue to the homebrew/core tap (not Homebrew/brew or Homebrew/core), or even better, submit a PR to fix it:
  /home/linuxbrew/.linuxbrew/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/graphviz.rb:12
```
To get around this, I've written an [external Homebrew command](https://docs.brew.sh/External-Commands) to extract and version the formula _without_ loading it into Homebrew. It's strictly a hack job, and may not do all the checks that `brew extract` does internally, but gets the job done for every formula I've tried:
```
brew tap gromgit/dev   # tap my dev repo
brew yank --help       # usage instructions

brew tap-new peteromueller/versions
brew yank graphviz 2.42.2 peteromueller/versions
brew install peteromueller/versions/graphviz@2.42.2
```

> How do I find out which older versions are supported / available?

`brew log graphviz` shows you the complete commit history for `graphviz`, and thanks to Homebrew's strict commit naming, it's easy to figure out which versions can be `extract`ed or `yank`ed.

As for supported versions, only the current version is ever supported. `brew extract/yank` comes with an implicit "you're on your own, buddy".

**Comment by peteromueller** [comment]
Hello,

I followed your suggestions and the result is excellent.
Now I measure around 100ms execution time again!!!

Libraries are as follows
        /opt/homebrew/Cellar/graphviz@2.42.2/2.42.2/lib/libgvc.6.dylib (compatibility version 7.0.0, current version 7.0.0)
        /opt/homebrew/opt/libtool/lib/libltdl.7.dylib (compatibility version 11.0.0, current version 11.1.0)
        /opt/homebrew/Cellar/graphviz@2.42.2/2.42.2/lib/libxdot.4.dylib (compatibility version 5.0.0, current version 5.0.0)
        /opt/homebrew/Cellar/graphviz@2.42.2/2.42.2/lib/libcgraph.6.dylib (compatibility version 7.0.0, current version 7.0.0)
        /opt/homebrew/Cellar/graphviz@2.42.2/2.42.2/lib/libpathplan.4.dylib (compatibility version 5.0.0, current version 5.0.0)
        /usr/lib/libexpat.1.dylib (compatibility version 7.0.0, current version 8.0.0)
        /usr/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.11)
        /usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1292.100.5)
        /opt/homebrew/Cellar/graphviz@2.42.2/2.42.2/lib/libcdt.5.dylib (compatibility version 6.0.0, current version 6.0.0)

People on graphviz group asked me if there are changes in the packaging of brew (not sure what they mean exactly).
Do you can think of something relevant for this problem here?
